arangod/Agency/Supervision.cpp

```diff
@@ -1079,6 +1108,163 @@ void Supervision::handleShutdown() {
+  { VPackObjectBuilder d(&del);
+    del.add("op", VPackValue("delete")); }
+
+  auto envelope = std::make_shared<VPackBuilder>();
+  VPackBuilder& trxs = *envelope; // Transactions
+  { VPackArrayBuilder t(&trxs);
+    for (auto const& server : servers) {
+      { VPackArrayBuilder ta(&trxs);
+        auto const serverName = server.first;
+        LOG_TOPIC("fa76d", DEBUG, Logger::SUPERVISION) <<
+          "Removing long overdue db server " << serverName << "last seen: " << server.second;
+        { VPackObjectBuilder oper(&trxs); // Operation for one server
+          trxs.add("/arango/Supervision/Health/" + serverName, del.slice());
+          trxs.add("/arango/Plan/DBServers/" + serverName, del.slice());
+          trxs.add("/arango/Current/DBServers/" + serverName, del.slice());
+          trxs.add("/arango/Target/MapUniqueToShortID/" + serverName, del.slice());
+          trxs.add("/arango/Current/ServersKnown/" + serverName, del.slice());
+          trxs.add("/arango/Current/ServersRegistered/" + serverName, del.slice());
+          for (auto const& j : currentDatabases.children()) {
+            trxs.add("/arango/Current/Databases/" + j.first + "/" + serverName, del.slice());
+          }
+        }
+        if (!server.second.empty()) { // Timestamp unchanged only, if persistent entry
+          VPackObjectBuilder prec(&trxs);
+          trxs.add("/arango/Supervision/Health/" + serverName + "/Timestamp", VPackValue(server.second));
+        }
+      }
+    }
+  }
+  if (servers.size() > 0) {
+    _nextServerCleanup = std::chrono::system_clock::now() + std::chrono::seconds(3600);
+    _agent->write(envelope);
+  }
+
+  trxs.clear();
+  servers = deletionCandidates(snapshot, transient, "Coordinators");
+  { VPackArrayBuilder t(&trxs);
+    for (auto const& server : servers) {
+      { VPackArrayBuilder ta(&trxs);
+        auto const serverName = server.first;
+        LOG_TOPIC("f6a7d", DEBUG, Logger::SUPERVISION) <<
+          "Removing long overdue coordinator " << serverName << "last seen: " << server.second;
+        { VPackObjectBuilder ts(&trxs);
+          trxs.add("/arango/Supervision/Health/" + serverName, del.slice());
+          trxs.add("/arango/Plan/Coordinators/" + serverName, del.slice());
+          trxs.add("/arango/Current/Coordinators/" + serverName, del.slice());
+          trxs.add("/arango/Target/MapUniqueToShortID/" + serverName, del.slice());
+          trxs.add("/arango/Current/ServersKnown/" + serverName, del.slice());
+          trxs.add("/arango/Current/ServersRegistered/" + serverName, del.slice());
+        }
+        if (!server.second.empty()) { // Timestamp unchanged only, if persistent entry
+          VPackObjectBuilder prec(&trxs);
+          trxs.add("/arango/Supervision/Health/" + serverName + "/Timestamp", VPackValue(server.second));
+        }
+      }
+    }
+  }
+  if (servers.size() > 0) {
+    _nextServerCleanup = std::chrono::system_clock::now() + std::chrono::seconds(3600);
```

Comment:
Should this not happen outside of the if? If nothing is deleted this looks like a busy look eventually.